packages/marketplace/src/components/ui/app-revision-comparision/app-revision-comparision.tsx

```diff
@@ -0,0 +1,161 @@
+import * as React from 'react'
+import { AppDetailState } from '@/reducers/app-detail'
+import { RevisionDetailState } from '@/reducers/revision-detail'
+import { AppRevisionModel, MediaModel, ScopeModel } from '@reapit/foundations-ts-definitions'
+import DiffMedia from '@/components/ui/diff-media'
+import DiffCheckbox from '../diff-checkbox'
+import DiffViewer from '../diff-viewer'
+
+export type AppRevisionComparisionProps = {
+  revisionDetailState: RevisionDetailState
+  appDetailState: AppDetailState
+}
+
+export type DiffMediaModel = {
+  currentMedia?: string
+  changedMedia?: string
+  order: number
+  type: string
+}
+
+const diffStringList: { [k in keyof AppRevisionModel]: string } = {
+  name: 'Name',
+  category: 'Category',
+  homePage: 'Home page',
+  launchUri: 'Launch URI',
+  supportEmail: 'Support Email',
+  telephone: 'Telephone',
+  summary: 'Summary',
+  description: 'Description',
+  redirectUris: 'Redirect URIs',
+  signoutUris: 'Signout URIs',
+}
+
+export const isAppearInScope = (nameNeedToFind: string | undefined, scopes: ScopeModel[] = []): boolean => {
+  if (!nameNeedToFind || scopes.length === 0) {
+    return false
+  }
+  const result = scopes.find((item: ScopeModel) => {
+    return item.name === nameNeedToFind
+  })
+  return !!result
+}
+
+export const renderCheckboxesDiff = ({
+  scopes,
+  appScopes,
+  revisionScopes,
+}: {
+  scopes: ScopeModel[]
+  appScopes: ScopeModel[] | undefined
+  revisionScopes: ScopeModel[] | undefined
+}) => {
+  return scopes.map((scope: ScopeModel) => {
+    const isCheckedInAppDetail = isAppearInScope(scope.name, appScopes)
+    const isCheckedInRevision = isAppearInScope(scope.name, revisionScopes)
+    return (
+      <div className="mb-3" key={scope.name}>
+        <h4 className="mb-2">{scope.description}</h4>
+        <DiffCheckbox currentChecked={isCheckedInAppDetail} changedChecked={isCheckedInRevision} />
+      </div>
+    )
+  })
+}
+
+export const getChangedMediaList = ({ app, revision }): DiffMediaModel[] => {
+  // Check the longest array to compare
+  const isNewMediaMoreItemThanOldOne = revision?.media?.length >= app?.media?.length
+  if (isNewMediaMoreItemThanOldOne) {
+    return revision.media.map((revisionMedia: MediaModel, index: number) => ({
+      changedMedia: revisionMedia?.uri,
+      currentMedia: app.media?.[index]?.uri,
+      order: revisionMedia?.order || 0,
+      type: revisionMedia?.type || '',
+    }))
+  }
+
+  return app.media.map((currentMedia: MediaModel, index: number) => ({
+    changedMedia: revision.media?.[index]?.uri,
+    currentMedia: currentMedia?.uri,
+    order: currentMedia?.order || 0,
+    type: currentMedia?.type || 'media',
+  }))
```

Comment:
Do we need to use `?` here?
if `revision.media` or `app.media`  is `undefined`, this function will return `undefined`, which not match `DiffMediaModel[]`. Should check manually rather than using opt chain`?`